File: app/src/main/java/com/boardgamegeek/provider/GamesIdCategoriesIdProvider.java
package com.boardgamegeek.provider;

import android.content.ContentUris;
import android.net.Uri;

import com.boardgamegeek.provider.BggContract.Categories;
import com.boardgamegeek.provider.BggDatabase.GamesCategories;
import com.boardgamegeek.util.SelectionBuilder;

public class GamesIdCategoriesIdProvider extends BaseProvider {
	GamesIdCategoriesProvider mProvider = new GamesIdCategoriesProvider();

	@Override
	protected SelectionBuilder buildSimpleSelection(Uri uri) {
		long categoryId = ContentUris.parseId(uri);
		return mProvider.buildSimpleSelection(uri).whereEquals(GamesCategories.CATEGORY_ID, categoryId);
	}

	@Override
	protected String getPath() {
		return addIdToPath(mProvider.getPath());
	}

	@Override
	protected String getType(Uri uri) {
		return Categories.CONTENT_ITEM_TYPE;
	}
}

File: app/src/main/java/com/boardgamegeek/provider/GamesIdCategoriesIdProvider.kt
package com.boardgamegeek.provider

import android.content.ContentUris
import android.net.Uri
import com.boardgamegeek.provider.BggContract.*
import com.boardgamegeek.provider.BggDatabase.GamesCategories.CATEGORY_ID
import com.boardgamegeek.provider.BggDatabase.GamesCategories.GAME_ID
import com.boardgamegeek.provider.BggDatabase.Tables
import com.boardgamegeek.util.SelectionBuilder

class GamesIdCategoriesIdProvider : BaseProvider() {
    override fun getType(uri: Uri) = Categories.CONTENT_ITEM_TYPE

    override val path = "$PATH_GAMES/#/$PATH_CATEGORIES/#"

    override fun buildSimpleSelection(uri: Uri): SelectionBuilder {
        val gameId = Games.getGameId(uri)
        val categoryId = ContentUris.parseId(uri)
        return SelectionBuilder()
                .table(Tables.GAMES_CATEGORIES)
                .whereEquals(GAME_ID, gameId)
                .whereEquals(CATEGORY_ID, categoryId)
    }
}
